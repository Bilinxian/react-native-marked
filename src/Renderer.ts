import { createElement } from 'react';
import { Linking, StyleSheet } from 'react-native';
import { Text } from 'react-native';
import { generateRandomID } from './utils';

const onLinkPress = (href: string) => () => Linking.openURL(href);

class Renderer {
  text(text: string) {
    const key = generateRandomID();
    return createElement(
      Text,
      {
        key,
        style: [styles.text],
      },
      text
    );
  }

  em(text: string) {
    const key = generateRandomID();
    return createElement(
      Text,
      {
        key,
        style: [styles.em, styles.text],
      },
      text
    );
  }

  link(text: string, href: string) {
    const key = generateRandomID();
    return createElement(
      Text,
      {
        key,
        style: [styles.em, styles.text, styles.link],
        onPress: onLinkPress(href),
      },
      text
    );
  }

  paragraph(children: React.ReactNode[]) {
    const key = generateRandomID();
    return createElement(
      Text,
      {
        key,
        style: [styles.text, styles.paragraph],
      },
      children
    );
  }
}

const styles = StyleSheet.create({
  em: {
    fontStyle: 'italic',
  },
  text: {
    fontSize: 14,
  },
  paragraph: {
    padding: 8,
  },
  link: {
    color: '#0074cc',
  },
});

export default Renderer;
