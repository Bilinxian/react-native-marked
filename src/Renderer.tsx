import * as React from 'react';
import { Linking, StyleSheet } from 'react-native';
import { Text } from 'react-native';
import { generateRandomString } from './utils';

class Renderer {
  private onLinkPress = (href: string) => () => Linking.openURL(href);

  text(text: string) {
    return (
      <Text key={generateRandomString()} style={[styles.text]}>
        {text}
      </Text>
    );
  }

  em(text: string) {
    return (
      <Text key={generateRandomString()} style={[styles.em, styles.text]}>
        {text}
      </Text>
    );
  }

  link(text: string, href: string) {
    return (
      <Text
        key={generateRandomString()}
        onPress={this.onLinkPress(href)}
        style={[styles.em, styles.text, styles.link]}
      >
        {text}
      </Text>
    );
  }

  paragraph(children: React.ReactNode[]) {
    return (
      <Text
        key={generateRandomString()}
        style={[styles.text, styles.paragraph]}
      >
        {children}
      </Text>
    );
  }
}

const styles = StyleSheet.create({
  em: {
    fontStyle: 'italic',
  },
  text: {
    fontSize: 14,
  },
  paragraph: {
    padding: 8,
  },
  link: {
    color: '#0074cc',
  },
});

export default Renderer;
