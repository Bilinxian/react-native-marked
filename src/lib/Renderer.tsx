import React, { ReactNode } from 'react';
import {
  ScrollView,
  View,
  Text,
  TouchableHighlight,
  TextStyle,
  ViewStyle,
} from 'react-native';
import MarkedList from '@jsamr/react-native-li';
import Disc from '@jsamr/counter-style/presets/disc';
import Decimal from '@jsamr/counter-style/presets/decimal';
import MDImage from './../components/MDImage';
import { generateRandomString } from '../utils/string';
import { onLinkPress } from '../utils/handlers';

class Renderer {
  getTextNode = (children: string | ReactNode[], styles?: TextStyle) => {
    return (
      <Text key={generateRandomString()} style={styles}>
        {children}
      </Text>
    );
  };

  getTextLinkNode = (
    children: string | ReactNode[],
    href: string,
    styles?: TextStyle
  ) => {
    return (
      <Text
        accessibilityRole="link"
        accessibilityHint="Opens in a new window"
        key={generateRandomString()}
        onPress={onLinkPress(href)}
        style={styles}
      >
        {children}
      </Text>
    );
  };

  getImageLinkNode = (href: string, imageUrl: string, alt?: string) => {
    const imageNode = this.getImageNode(imageUrl, alt);
    return (
      <TouchableHighlight
        accessibilityRole="link"
        accessibilityHint="Opens in a new window"
        onPress={onLinkPress(href)}
        key={generateRandomString()}
      >
        {imageNode}
      </TouchableHighlight>
    );
  };

  getViewNode(children: ReactNode[] | null, styles?: ViewStyle) {
    return (
      <View key={generateRandomString()} style={styles}>
        {children}
      </View>
    );
  }

  getCodeBlockNode(
    text: string,
    containerStyle?: ViewStyle,
    textStyle?: TextStyle
  ) {
    return (
      <ScrollView
        key={generateRandomString()}
        contentContainerStyle={containerStyle}
        horizontal
      >
        <Text style={textStyle}>{text}</Text>
      </ScrollView>
    );
  }

  getBlockquoteNode(children: ReactNode[], styles?: ViewStyle) {
    return (
      <View key={generateRandomString()} style={styles}>
        {children}
      </View>
    );
  }

  getImageNode(uri: string, alt?: string) {
    return <MDImage key={generateRandomString()} uri={uri} alt={alt} />;
  }

  getListNode(
    ordered: boolean,
    li: ReactNode[],
    listStyle?: ViewStyle,
    textStyle?: TextStyle
  ) {
    return (
      <MarkedList
        counterRenderer={ordered ? Decimal : Disc}
        markerTextStyle={textStyle}
        markerBoxStyle={listStyle}
        key={generateRandomString()}
      >
        {li.map((node) => node)}
      </MarkedList>
    );
  }
}

export default Renderer;
