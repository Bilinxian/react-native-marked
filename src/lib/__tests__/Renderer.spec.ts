import { fireEvent, render, screen } from '@testing-library/react-native';
import Renderer from '../Renderer';
import getStyles from './../../theme/styles';

const styles = getStyles();
const renderer = new Renderer();

describe('Renderer', () => {
  describe('getTextNode', () => {
    it('returns a Text node', () => {
      const TextNode = renderer.getTextNode('Hello world', styles.text);

      const r = render(TextNode);
      expect(screen.queryByText('Hello world')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('returns a wrapped Text node', () => {
      const TextNodeChild = renderer.getTextNode('Hello world', {});
      const TextNode = renderer.getTextNode([TextNodeChild], styles.text);
      const r = render(TextNode);
      expect(screen.queryByText('Hello world')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('returns a wrapped Text node with styles', () => {
      const TextNodeChild = renderer.getTextNode('Hello world', styles.text);
      const TextNode = renderer.getTextNode([TextNodeChild], styles.text);
      const r = render(TextNode);
      expect(screen.queryByText('Hello world')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('getLinkNode', () => {
    it('returns a Link node', () => {
      const LinkNode = renderer.getLinkNode(
        'Link',
        'https://example.com',
        styles.link
      );
      const r = render(LinkNode);
      expect(screen.queryByText('Link')).toBeTruthy();
      fireEvent.press(screen.queryByText('Link'));
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('getViewNode', () => {
    it('returns a paragraph View node', () => {
      const TextNode = renderer.getTextNode('Hello world', styles.text);
      const LinkNode = renderer.getLinkNode(
        'Link',
        'https://example.com',
        styles.link
      );
      const ViewNode = renderer.getViewNode(
        [TextNode, LinkNode],
        styles.paragraph
      );

      const r = render(ViewNode);
      expect(screen.queryByText('Hello world')).toBeTruthy();
      expect(screen.queryByText('Link')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('returns a hr View node', () => {
      const ViewNode = renderer.getViewNode(null, styles.hr);
      const r = render(ViewNode);
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('getCodeBlockNode', () => {
    it('returns a Code block (horizontal ScrollView)', () => {
      const CodeBlock = renderer.getCodeBlockNode(
        "print('hello')",
        styles.code,
        styles.em
      );
      const r = render(CodeBlock);
      expect(screen.queryByText("print('hello')")).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('getBlockquoteNode', () => {
    it('returns a Blockquote', () => {
      const TextNode = renderer.getTextNode('Hello world', styles.text);
      const LinkNode = renderer.getLinkNode(
        'Link',
        'https://example.com',
        styles.link
      );
      const Blockquote = renderer.getBlockquoteNode(
        [TextNode, LinkNode],
        styles.blockquote
      );

      const r = render(Blockquote);
      expect(screen.queryByText('Hello world')).toBeTruthy();
      expect(screen.queryByText('Link')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('getImageNode', () => {
    it('returns a Image', () => {
      const ImageNode = renderer.getImageNode('https://picsum.photos/100/100');
      const tree = render(ImageNode).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('getListNode', () => {
    it('returns Ordered List', () => {
      const TextNode1 = renderer.getTextNode('Hello world 1', styles.li);
      const TextNode2 = renderer.getTextNode('Hello world 2', styles.li);
      const TextNode3 = renderer.getTextNode('Hello world 3', styles.li);
      const OL = renderer.getListNode(
        true,
        [TextNode1, TextNode2, TextNode3],
        styles.list,
        styles.li
      );
      const r = render(OL);
      expect(screen.queryByText('Hello world 1')).toBeTruthy();
      expect(screen.queryByText('Hello world 2')).toBeTruthy();
      expect(screen.queryByText('Hello world 3')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('returns Un-Ordered List', () => {
      const TextNode1 = renderer.getTextNode('Hello world 1', styles.li);
      const TextNode2 = renderer.getTextNode('Hello world 2', styles.li);
      const TextNode3 = renderer.getTextNode('Hello world 3', styles.li);
      const OL = renderer.getListNode(
        false,
        [TextNode1, TextNode2, TextNode3],
        styles.list,
        styles.li
      );
      const r = render(OL);
      expect(screen.queryByText('Hello world 1')).toBeTruthy();
      expect(screen.queryByText('Hello world 2')).toBeTruthy();
      expect(screen.queryByText('Hello world 3')).toBeTruthy();
      const tree = r.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
