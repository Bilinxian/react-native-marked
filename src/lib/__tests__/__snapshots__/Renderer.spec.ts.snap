// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer getBlockquoteNode returns a Blockquote 1`] = `
<View
  style={
    {
      "borderLeftColor": "#d0d7de",
      "borderLeftWidth": 5,
      "opacity": 0.8,
      "paddingLeft": 16,
    }
  }
>
  <Text
    style={
      {
        "color": "#333333",
        "fontSize": 16,
        "lineHeight": 24,
      }
    }
  >
    Hello world
  </Text>
  <Text
    onPress={[Function]}
    style={
      {
        "color": "#58a6ff",
        "fontSize": 16,
        "fontStyle": "italic",
        "lineHeight": 24,
      }
    }
  >
    Link
  </Text>
</View>
`;

exports[`Renderer getCodeBlockNode returns a Code block (horizontal ScrollView) 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "backgroundColor": "#f6f8fa",
      "minWidth": "100%",
      "padding": 16,
    }
  }
  horizontal={true}
>
  <View>
    <Text
      style={
        {
          "color": "#333333",
          "fontSize": 16,
          "fontStyle": "italic",
          "lineHeight": 24,
        }
      }
    >
      print('hello')
    </Text>
  </View>
</RCTScrollView>
`;

exports[`Renderer getImageNode returns a Image 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    [
      undefined,
      {},
      {
        "height": 0,
      },
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
    ]
  }
>
  <Image
    onError={[Function]}
    onLayout={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    resizeMode="cover"
    source={
      {
        "uri": "https://picsum.photos/100/100",
      }
    }
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 0,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Renderer getLinkNode returns a Link node 1`] = `
<Text
  onPress={[Function]}
  style={
    {
      "color": "#58a6ff",
      "fontSize": 16,
      "fontStyle": "italic",
      "lineHeight": 24,
    }
  }
>
  Link
</Text>
`;

exports[`Renderer getListNode returns Ordered List 1`] = `
[
  <View
    style={
      [
        {
          "alignSelf": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        undefined,
      ]
    }
    testID="marked-list-item"
  >
    <View>
      <Text
        style={
          [
            {
              "alignSelf": "flex-end",
              "color": "#333333",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontSize": 16,
              "lineHeight": 24,
              "textAlign": "right",
            },
            {
              "width": 28.799999999999997,
            },
          ]
        }
        testID="marker-box"
      >
        1. 
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#333333",
          "flexShrink": 1,
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      Hello world 1
    </Text>
  </View>,
  <View
    style={
      [
        {
          "alignSelf": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        undefined,
      ]
    }
    testID="marked-list-item"
  >
    <View>
      <Text
        style={
          [
            {
              "alignSelf": "flex-end",
              "color": "#333333",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontSize": 16,
              "lineHeight": 24,
              "textAlign": "right",
            },
            {
              "width": 28.799999999999997,
            },
          ]
        }
        testID="marker-box"
      >
        2. 
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#333333",
          "flexShrink": 1,
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      Hello world 2
    </Text>
  </View>,
  <View
    style={
      [
        {
          "alignSelf": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        undefined,
      ]
    }
    testID="marked-list-item"
  >
    <View>
      <Text
        style={
          [
            {
              "alignSelf": "flex-end",
              "color": "#333333",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontSize": 16,
              "lineHeight": 24,
              "textAlign": "right",
            },
            {
              "width": 28.799999999999997,
            },
          ]
        }
        testID="marker-box"
      >
        3. 
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#333333",
          "flexShrink": 1,
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      Hello world 3
    </Text>
  </View>,
]
`;

exports[`Renderer getListNode returns Un-Ordered List 1`] = `
[
  <View
    style={
      [
        {
          "alignSelf": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        undefined,
      ]
    }
    testID="marked-list-item"
  >
    <View>
      <Text
        style={
          [
            {
              "alignSelf": "flex-end",
              "color": "#333333",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontSize": 16,
              "lineHeight": 24,
              "textAlign": "right",
            },
            {
              "width": 19.2,
            },
          ]
        }
        testID="marker-box"
      >
        • 
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#333333",
          "flexShrink": 1,
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      Hello world 1
    </Text>
  </View>,
  <View
    style={
      [
        {
          "alignSelf": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        undefined,
      ]
    }
    testID="marked-list-item"
  >
    <View>
      <Text
        style={
          [
            {
              "alignSelf": "flex-end",
              "color": "#333333",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontSize": 16,
              "lineHeight": 24,
              "textAlign": "right",
            },
            {
              "width": 19.2,
            },
          ]
        }
        testID="marker-box"
      >
        • 
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#333333",
          "flexShrink": 1,
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      Hello world 2
    </Text>
  </View>,
  <View
    style={
      [
        {
          "alignSelf": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
        },
        undefined,
      ]
    }
    testID="marked-list-item"
  >
    <View>
      <Text
        style={
          [
            {
              "alignSelf": "flex-end",
              "color": "#333333",
              "flexGrow": 0,
              "flexShrink": 1,
              "fontSize": 16,
              "lineHeight": 24,
              "textAlign": "right",
            },
            {
              "width": 19.2,
            },
          ]
        }
        testID="marker-box"
      >
        • 
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#333333",
          "flexShrink": 1,
          "fontSize": 16,
          "lineHeight": 24,
        }
      }
    >
      Hello world 3
    </Text>
  </View>,
]
`;

exports[`Renderer getTextNode returns a Text node 1`] = `
<Text
  style={
    {
      "color": "#333333",
      "fontSize": 16,
      "lineHeight": 24,
    }
  }
>
  Hello world
</Text>
`;

exports[`Renderer getTextNode returns a wrapped Text node 1`] = `
<Text
  style={
    {
      "color": "#333333",
      "fontSize": 16,
      "lineHeight": 24,
    }
  }
>
  <Text
    style={{}}
  >
    Hello world
  </Text>
</Text>
`;

exports[`Renderer getTextNode returns a wrapped Text node with styles 1`] = `
<Text
  style={
    {
      "color": "#333333",
      "fontSize": 16,
      "lineHeight": 24,
    }
  }
>
  <Text
    style={
      {
        "color": "#333333",
        "fontSize": 16,
        "lineHeight": 24,
      }
    }
  >
    Hello world
  </Text>
</Text>
`;

exports[`Renderer getViewNode returns a hr View node 1`] = `
<View
  style={
    {
      "borderBottomColor": "#d0d7de",
      "borderBottomWidth": 1,
      "marginVertical": 4,
    }
  }
/>
`;

exports[`Renderer getViewNode returns a paragraph View node 1`] = `
<View
  style={
    {
      "color": "#333333",
      "fontSize": 16,
      "lineHeight": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#333333",
        "fontSize": 16,
        "lineHeight": 24,
      }
    }
  >
    Hello world
  </Text>
  <Text
    onPress={[Function]}
    style={
      {
        "color": "#58a6ff",
        "fontSize": 16,
        "fontStyle": "italic",
        "lineHeight": 24,
      }
    }
  >
    Link
  </Text>
</View>
`;
